### DISCUSSION TITLE
How can I deploy to a solution outside of Streamlit Community Cloud and use Firestore functionality?

### DISCUSSION THREAD

**Discussion by emigre459** [444B/streamlit-analytics2#126] [Q&A]
Right now the SA2 package requires your service account JSON key to be stored in `.streamlit/secrets.toml` but I expect that will fail when I deploy my downstream SA2-enabled app to GCP Cloud Run (because I don't want to send secrets as a file in the image build). Is there a deployment pattern that would allow me to use GCP Secrets Manager/the default GCP credentials (since that's where my Firestore instance is located) without modifying the SA2 package, or is this something we should work out via PR?

**Comment by 444B** [comment]
Hellloo again
You can use secrets manager + environment variables in Cloud Run, that should work just fine
https://cloud.google.com/run/docs/configuring/services/environment-variables
I was able to make this work before on cloud run... Let me find my old code 

FYI as a firebase user, I need your input on #127 please

**Reply by 444B** [reply]
Hmmm, older wiser me does not approve
`streamlit_analytics.start_tracking(firestore_key_file="firebase-key.json", firestore_collection_name="inksight-counts")`

Here was my Arch FYI so if you have any GCP questions feel free as I have alot of experience
![313693172-e43e49e0-1ad9-4023-b779-d78b912fae68](https://github.com/user-attachments/assets/6153f8c4-1543-4b01-8973-7db309ed7a6b)

It was pretty effective, it even withstood a DoS of 27 million requests in under 24 hours and cost me nothing because of Cloud Armor - priceless

There is no ways I was throwing the key into the container. I think I was granting access to the key with this https://cloud.google.com/run/docs/configuring/services/secrets#access-secrets

If I was, then I ask the Devops Gods for forgiveness 🙏

**Reply by 444B** [reply]
I have a feeling that using an ENV variable in code and exposing with Cloud Run would be sufficient

**Reply by 444B** [reply]
Hang on, why wouldn't the st.secrets work outside of community cloud?
Am I missing something? You can just make a .streamlit/secrets.toml file and keep the key there
Or are you concerned of having it in text form anywhere in the container?

There must be a way with secret manager to get the json

**Reply by emigre459** [reply]
I'm concerned with having the text in the container for a long-running app as that would potentially be accessible by third parties (unlikely admittedly, but I'm less comfortable with that approach than with simply using env vars setup by Secrets Manager for it). I have a hacky solution with env vars now, but seems like it would be good to enable usage of GCP cloud auth native solutions (e.g. in my custom GCP clients I often use:

```python
def get_credentials() -> google.auth.credentials.Credentials:
    try:
        credentials, project_id = google.auth.default()
    except DefaultCredentialsError:
        credentials = None
        logger.warning("Failed to get default credentials", exc_info=True)
    return credentials
```

and have my wrapper clients call this function if they are instantiated with `credentials=None`.

**Reply by pvpiv** [reply]
Curious as well - I've been running the sa package until last month. Updated reference to sa2 and things were ok. Then the counts was switched to data and I had to update my code (still using custom load and save functions cause I have my firestore.json file in st.secrets

```python
def load_from_firestore(counts, collection_name):
    key_dict = json.loads(st.secrets["textkey"])
    creds = service_account.Credentials.from_service_account_info(key_dict)
    db = firestore.Client(credentials=creds, project="pvpogo")
    col = db.collection(collection_name)
    firestore_counts = col.document(st.secrets["fb_col"]).get().to_dict()
    if firestore_counts is not None:
        for key in firestore_counts:
            if key in counts:
                counts[key] = firestore_counts[key]

def save_to_firestore(counts, collection_name):
    key_dict = json.loads(st.secrets["textkey"])
    creds = service_account.Credentials.from_service_account_info(key_dict)
    db = firestore.Client(credentials=creds, project="pvpogo")
    col = db.collection(collection_name)
    col.document(st.secrets["fb_col"]).set(counts)
```


I'm a noob and I don't know how to access the file from community cloud so I took this route.

What is your recommendation @444B for me going forward? I want to rely on the package as is deployed (without having to edit my code for every update). Thank you for your service sir!

**Reply by 444B** [reply]
@pvpiv Thank you for sharing!  
First of all I want to apologize if the recent release 0.10.0 had an unintended side effect when we renamed the `counts` to `data`  
[We take steps ](https://github.com/444B/streamlit-analytics2/pull/127#issuecomment-2585847192)to consider any changed behaviour and how it may impact your use of the package  
Although we took those steps, I actually think the change that you are describing comes come this change [here](https://github.com/444B/streamlit-analytics2/pull/127/files#diff-9b4a3630806b652e2c3f78ff1ad799969576dc37044d33a271f999af56e7ddb3) since it appears you may have been using the `counts` variable directly. Did I understand correctly?  


I strive to keep no breaking changes between SA and SA2 while also trying to fix certain issues and add features.  

At a certain point I will need to induce a breaking change however that would be made aware to users before the release and they can make a choice  

For now, I will see if I can go to the init.py file and get it working with `from .state import data as counts` alongside the current implementation so that you will not need to change your application  

I fully intend to mimimize any need for you to change your code unless there is a major version release. We use [semantic versioning](https://semver.org/), so it will be possible for you to know if there is any breaking changes.  

TLDR; 
- I am sorry for any impact or unintended changes
- I will try fix it in a 0.10.3 release to expose the counts, alongside data if you use it directly in your app. I hadnt considered that some users might be.  
- at some point we will need to break some things if we want to get this package more useful and you will be made fully aware of the changes ahead of time :)

**Reply by pvpiv** [reply]
@444B Oh no worries, thank YOU for the continued support. And ye you got it right and I already made the change in my code. Was as simple - just changed .counts to .data in my function call and it all worked out perfectly.

**Reply by 444B** [reply]
Thank you again for pointing that out. 
This behaviour will maintain backwards compatibility in the [latest commit](https://github.com/444B/streamlit-analytics2/commit/a6233840de6605ae239caa07690046c81ccf70ce) towards 0.10.3

I will give some thought on how to tackle the longer term issues with deploying the firestore keys in community cloud and other places  

Personally I think parr of the issue is that the only way to save SA2 Data is in a JSON.
I want to allow people options to save and load the data from relational storage types such as CSV, sqlite3.   

I have some thoughts on this for the 0.11.0 or later release, let me know if youre interested and would like to be involved in early testing (the beta packages can be placed on testpypi)

**Comment by emigre459** [comment]
In case it's useful to anyone on here, here's our hacky solution for managing the secrets a little more securely through env vars (as other approaches that seemed more straightforward didn't work for us - this is a function at the top of our main app page that is called as soon as streamlit spins up):

```python
def setup_streamlit_secrets(env_var_name: str = "firebase"):
    # Check that secret isn't already set
    if st.secrets._secrets is not None and env_var_name in st.secrets._secrets.keys():
        logger.debug("Secret '%s' already set in st.secrets.", env_var_name)
        return
    elif st.secrets._secrets is None:
        st.secrets._secrets = {}
    st.secrets._secrets[env_var_name] = os.getenv(env_var_name)

setup_streamlit_secrets()
```
